lighthouse-core/audits/no-unload-listeners.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const i18n = require('./../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Descriptive title of a Lighthouse audit that checks if a web page has 'unload' event listeners and finds none.  */
+  title: 'Avoids `unload` event listeners',
+  /** Descriptive title of a Lighthouse audit that checks if a web page has 'unload' event listeners and finds that it is using them. */
+  failureTitle: 'Listens for the `unload` event',
+  /** Description of a Lighthouse audit that tells the user why pages should not use the 'unload' event. This is displayed after a user expands the section to see more. 'Learn More' becomes link text to additional documentation. */
+  description: 'The `unload` event does not fire reliably and listening for it can prevent browser optimizations like the back/forward cache. Consider using the `pagehide` or `visibilitychange` events instead. [Learn More](https://developers.google.com/web/updates/2018/07/page-lifecycle-api#legacy-lifecycle-apis-to-avoid)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class NoUnloadListeners extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'no-unload-listeners',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['UnloadListeners', 'JsUsage'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {LH.Audit.Product}
+   */
+  static audit(artifacts) {
+    const unloadListeners = artifacts.UnloadListeners.filter(l => l.type === 'unload');
+    if (!unloadListeners.length) {
+      return {
+        score: 1,
+      };
+    }
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headings = [
+      {key: 'url', itemType: 'url', text: str_(i18n.UIStrings.columnURL)},
+      {key: 'line', itemType: 'text', text: str_(i18n.UIStrings.columnLocation)},
+    ];
+
+    const jsUsageValues = Object.values(artifacts.JsUsage)
```

Comment:
any reason to not just go for a map at this point if you've already got to do some processing? :)

Reply:
> any reason to not just go for a map at this point if you've already got to do some processing? :)

ha, well I was going to, but it also seemed like excessive work to create a Map for all the scriptIds when we don't care about deduping, etc, just whatever the first entry is for 1 or 2 scriptIds, but also happy to make it a map if it reads more easily.

Reply:
I'm on team Map for sure :)

confusing to read it flatten the array only to then use it like a map 😆